### ISSUE TITLE
Refactor locate service code

### ISSUE THREAD

**Issue by jaredkerim** [mozilla/ichnaea#400]
The code located at:

https://github.com/mozilla/ichnaea/blob/master/ichnaea/service/locate.py

Is designed partially to allow an extensible set of search types, but not in a very clean and modular way.  We should refactor this code to break the search logic into two parts: a SearchBroker, and a SearchProvider.  

A SearchProvider is responsible for performing one specific type of search, so for our current implementation that would be each of: cell, OCID cell, cell lac, wifi, geoip.  In the future we will add OCID cell lac (#351), as well as external API calls.  To prepare for that, refactoring the code now will make extending the number and types of search providers easier.

The SearchBroker will collect the search results of each SearchProvider, and either aggregate or decide which to use.  Presently the search logic uses a 'drill down' approach from the broadest to the narrowest search type and uses the best match it gets, if any.  When the SearchBroker and SearchProvider are clearly distinct form one another, changing how this logic behaves will be simpler without having to modify how each SearchProvider implements its specific type of search.

**Comment by hannosch** [comment]
This has been merged and is all done. I've expanded a bit on it, unifying the geoip and database session based providers and introducing a stateful location result object. The result object made it possible to simplify the main loop by doing things like "new_result.more_accurate(old_result)", moving the logic on how to compare results onto these result objects.

In order to fix the last TODO (cell/cell_lac metrics) I had to add back some rather ugly code. This is going to get even more messy when we add the ocid_area provider. Maybe some other pattern is going to be obvious once we do that.

The overall design has already proven itself, as I managed to almost trivially add a new CellCountryProvider for providing cell key (mcc) based country estimates.

**Comment by hannosch** [comment]
Actually there's a bit more general refactoring that could be done, that isn't specific to #351.

To make the logging work, I had to introduce a log_group name on each provider and do some grouped stats gathering at the end of the main locate loop. Maybe this could be generalized into the concept of a provider group, with one group per incoming data field. So we'd loop over three groups (geoip, cell, wifi) and one to multiple providers inside each group. That might make it more obvious to distinguish between a "cell_miss", "cell_hit" and "cell_lac_hit", as we can only determine that after we queried all cell based providers.

I've also started to factor out the concept of result quality / comparison into location result classes. It's already helpful to give GeoIP a general lower trust value (based on a class based lower priority value). But I think there's another pass we could do to make this more explicit. In particular let each provider return multiple location results (for example wifi might result in two clusters, a multi-sim device might get a gsm and umts cell estimate). 

Today each provider has a bit of logic inside it to arrive at "the best guess" for its own data (based on wifi signals and number of cell stations). But then we have another part which wants to match the results from different providers against each other and only use them if they match (lat/lon is inside the lat/lon/accuracy range of one another).

Currently this means we might get two wifi clusters, the first being bogus (some outdated or yet undetected moving wifis). But this bogus result is the only one being returned by the wifi provider. The cell provider might give us correct data and upon comparing it with the bogus wifi result we discard the wifi result. If we let each provider return multiple results with some sort of score, we could choose to use wifi cluster number two, which can be be confirmed by the cell data. Overall resulting in a more accurate wifi based estimate.

The third idea in the code is some sort of "accurate enough" score. Ideally we'd reverse the lookup order and go from wifi, cell to geoip. If we get a wifi result being confirmed by our own cell table, there's no need for us to query any of the other cell area or ocid cell / area tables. This gets more important once we introduce an outside fallback service, with non-trivial latency costs. We need some scoring to decide whether or not we trust our own data enough, or whether we should consult an outside fallback provider.
